### PR TITLE
Disambiguating Zuora subscription data from Membership vs Subscriptions

### PR DESCRIPTION
**Pull request by afiore**
_PR description by @rtyley :_

These are some fairly intense changes to the `SubscriptionService` to ensure that it correctly disambiguates between Zuora data for Membership and for Subscriptions (now the Digital Pack 2.0 is soon to launch).

The key thing to know about Zuora data:

* **Zuora accounts** do **not** directly relate to Membership vs Subscriptions (you can not say 'this Zuora Account is a Membership Account'). A Zuora account only has to exist to provide a default payment method - so if a user is only paying with credit card, they should only have one account, but if they have both direct debit AND credit card, they will need _two_ Zuora accounts. PayPal, etc, would require a third.
* **Zuora Subscriptions** are only _loosely_ associated with Membership or Subscriptions. You could have a Zuora Subscription which had both membership.theguardian.com & subscribe.theguardian.com products on it - we don't currently create those, but there is no guarantee that they won't occur in future.
* **Zuora Products** _are_ associated with either Membership or Subscriptions. A Product might be 'Supporter' membership, or 'Digital Pack' for subscribe.theguardian.com.

So when it comes to working out which _Zuora Subscription_ you want to work with, you have to see which Products it has. If it has Membership Products, then it's a Membership subscription. The Zuora account that the subscription falls under could then be considered a 'Membership' account for some purposes - but might also include Zuora Subscriptions for subscribe.theguardian.com.

We are taking advantage of the Guardian custom field `ProductType` which already exists on Zuora Products. For Partner, Patron, Supporter, Staff Membership, this is correctly set to 'Membership'. For 'Digital Pack' the `ProductType` is 'Digital Pack' which I think may not be a good value for it - it should probably be 'Subscriptions'.

cc @tudorraul @vivekkr 

https://trello.com/c/ltEhCsjV/37-filter-non-membership-zoura-subscriptions-to-avoid-clash-with-subscribe-theguardian-com

### CODE REVIEW

File: frontend/app/services/zuora/ZuoraRestService.scala [hunk 1]
```diff
@@ -32,12 +33,16 @@ class ZuoraRestService(config: ZuoraApiConfig) extends LazyLogging {
     }
   }
 
+  val productCatalogSupplier = new FutureSupplier[ProductCatalog](productCatalog)
```
Comment (rtyley) [inline frontend/app/services/zuora/ZuoraRestService.scala]: @vivekkr @tudorraul this is where you can start grabbing the Zuora Product Catalog now

File: frontend/app/model/Zuora.scala [hunk 2]
```diff
@@ -55,8 +55,10 @@ object Zuora {
   }
   case class InternalError(code: String, message: String) extends Error
 
-  case class SubscriptionStatus(current: String, future: Option[String], amendType: Option[String]) {
-    val cancelled = amendType.exists(_ == "Cancellation")
+  case class SubscriptionStatus(current: Subscription, future: Option[Subscription], amendType: Option[String]) {
```
Comment (tudorraul) [inline frontend/app/model/Zuora.scala]: This looks to suggest a subscription _change_ rather than a subscription status.

Comment (tudorraul) [inline frontend/app/model/Zuora.scala]: Also, the wording here can be a bit confusing, `future` suggests a Scala `Future`. Perhaps `newVersion` might hold the meaning better.

File: frontend/app/services/SubscriptionService.scala [hunk 3]
```diff
@@ -105,76 +106,74 @@ object SubscriptionService {
...
+    amendments <- zuoraSoapService.query[Amendment](where)
+  } yield {
+      val latestSubscription = subscriptionVersions.maxBy(_.version)
+      sortAmendments(subscriptionVersions, amendments)
         .find(_.contractEffectiveDate.isAfterNow)
-        .fold(SubscriptionStatus(latestSubscriptionId, None, None)) { amendment =>
-          SubscriptionStatus(amendment.subscriptionId, Some(latestSubscriptionId), Some(amendment.amendType))
-        }
+        .fold(SubscriptionStatus(latestSubscription, None, None)) { futureAmendment =>
+          logger.info(s"subscription = ${subscription.id}, latestSubscriptionId ${latestSubscription.id}")
```
Comment (tudorraul) [inline frontend/app/services/SubscriptionService.scala]: what is this logging statement helping us with?

File: frontend/app/services/SubscriptionService.scala [hunk 4]
```diff
@@ -105,76 +106,74 @@ object SubscriptionService {
...
+  def memberTierFeatures(memberId: MemberId): Future[Seq[Rest.Feature]] = for {
+    (_, subscription) <- accountWithLatestMembershipSubscription(memberId)
+    catalog  <- zuoraRestService.productCatalogSupplier.get
+  } yield subscription
+      .latestWhiteListedRatePlan(catalog.productIdsOfType(MembershipProductType)).toSeq
+      .flatMap(_.subscriptionProductFeatures)
+
+  /**
+   * @return the current subscription version of the user, and the future subscription version, if
+   *         they have a pending downgrade.
```
Comment (tudorraul) [inline frontend/app/services/SubscriptionService.scala]: Does this method work only for downgrades, or does it for upgrades too?

File: frontend/app/services/SubscriptionService.scala [hunk 5]
```diff
@@ -105,76 +106,74 @@ object SubscriptionService {
...
+   */
+  def getSubscriptionStatus(memberId: MemberId): Future[SubscriptionStatus] = for {
+    (_, subscription) <- accountWithLatestMembershipSubscription(memberId)
+    subscriptionVersions <- subscriptionVersions(subscription.subscriptionNumber)
+    where = subscriptionVersions.map { sub => s"SubscriptionId='${sub.id}'" }.mkString(" OR ")
+    amendments <- zuoraSoapService.query[Amendment](where)
+  } yield {
+      val latestSubscription = subscriptionVersions.maxBy(_.version)
+      sortAmendments(subscriptionVersions, amendments)
```
Comment (tudorraul) [inline frontend/app/services/SubscriptionService.scala]: a bit unclear why sorting is necessary here?

File: frontend/app/services/SubscriptionService.scala [hunk 6]
```diff
@@ -105,76 +106,74 @@ object SubscriptionService {
...
-
-      where = subscriptions.map { sub => s"SubscriptionId='${sub.id}'" }.mkString(" OR ")
-      amendments <- zuoraSoapService.query[Amendment](where)
-    } yield {
-      val latestSubscriptionId = sortSubscriptions(subscriptions).last.id
+  import SubscriptionService._
+  val MembershipProductType = "Membership"
+
+  private def subscriptionVersions(subscriptionNumber: String): Future[Seq[Subscription]] = for {
+    subscriptions <- zuoraSoapService.query[Subscription](s"Name = '$subscriptionNumber'")
```
Comment (tudorraul) [inline frontend/app/services/SubscriptionService.scala]: no need for a `for` comprehension here

File: frontend/app/services/SubscriptionService.scala [hunk 7]
```diff
@@ -105,76 +106,76 @@ object SubscriptionService {
...
+  } yield subscriptions
+
+  def accountWithLatestMembershipSubscription(member: MemberId): Future[(Account, Rest.Subscription)] = for {
+    accounts <- zuoraSoapService.query[Account](s"crmId='${member.salesforceAccountId}'")
```
Comment (tudorraul) [inline frontend/app/services/SubscriptionService.scala]: I am in favour of concealing this ZQL logic in the `ZuoraSoapService`

File: frontend/app/services/zuora/Rest.scala [hunk 8]
```diff
@@ -31,6 +31,16 @@ object Rest {
   case object Cancelled extends SubscriptionStatus
   case object Expired extends SubscriptionStatus
 
+  /*
+   * Zuora product catalogue entry: https://knowledgecenter.zuora.com/BC_Developers/REST_API/B_REST_API_reference/Catalog
+   */
+  case class ProductCatalog(products: List[Product]) {
```
Comment (tudorraul) [inline frontend/app/services/zuora/Rest.scala]: should the return generic type be uniformed in these 2 methods? i.e. both `Seq` or bothe `Set`

File: frontend/app/services/SubscriptionService.scala (hunk 7 above)
Comment (afiore) [inline frontend/app/services/SubscriptionService.scala]: Agreed, but out of the scope of this PR really, as it is something we do in a number of places

Comment (afiore) [inline frontend/app/services/SubscriptionService.scala]: Also, notice that `query` and `queryOne` are pretty much the only public method in that class

File: frontend/app/services/SubscriptionService.scala (hunk 4 above)
Comment (afiore) [inline frontend/app/services/SubscriptionService.scala]: This currently works only for upgrades (which was the case also for the previous versions of this code). We plan on handling downgrades after we close this PR

**Comment by rtyley** [comment]
cc @TesterSpike

File: frontend/app/services/SubscriptionService.scala [hunk 9]
```diff
@@ -185,46 +186,44 @@ class SubscriptionService(val tierPlanRateIds: Map[ProductRatePlan, String],
...
   }
 
   def getMembershipSubscriptionSummary(memberId: MemberId): Future[MembershipSummary] = {
-
-    val latestSubF = {
-      for (subscriptions <- getSubscriptions(memberId))
-      yield sortSubscriptions(subscriptions).last
-    }
+    val latestSubF = for {
+      (_, subscription) <- accountWithLatestMembershipSubscription(memberId)
```
Comment (tudorraul) [inline frontend/app/services/SubscriptionService.scala]: There's quite a bit of redundancy happening behind the scenes here, and it's because we don't have a get subscription by account anymore. Creating one back and using that instead of the heavy `accountWithLatestMembershipSubscription` will cut back on this redundancy.

File: frontend/app/services/SubscriptionService.scala [hunk 10]
```diff
@@ -185,46 +186,44 @@ class SubscriptionService(val tierPlanRateIds: Map[ProductRatePlan, String],
...
-      } yield {
-        assert(result.invoiceItems.nonEmpty, "Subscription with delayed payment returning zero invoice items in SubscriptionDetailsViaAmend call")
-        val firstPreviewInvoice = result.invoiceItems.sortBy(_.serviceStartDate).head
-
-        MembershipSummary(latestSubscription.termStartDate, firstPreviewInvoice.serviceEndDate, None,
-          subscriptionDetails.planAmount, firstPreviewInvoice.price, firstPreviewInvoice.serviceStartDate, firstPreviewInvoice.renewalDate )
-      }
+    def getSummaryViaSubscriptionAmend(memberId: MemberId) = for {
+      latestSubscription <- latestSubF
+      subscriptionDetailsF = getSubscriptionDetails(latestSubscription)
```
Comment (tudorraul) [inline frontend/app/services/SubscriptionService.scala]: why the indirection?

Comment (tudorraul) [inline frontend/app/services/SubscriptionService.scala]: why the indirection?

File: frontend/app/services/zuora/ZuoraRestService.scala (hunk 1 above)
Comment (tudorraul) [inline frontend/app/services/zuora/ZuoraRestService.scala]: :+1:

File: frontend/app/services/SubscriptionService.scala (hunk 7 above)
Comment (chrisjowen) [inline frontend/app/services/SubscriptionService.scala]: Confusion for me comes form the fact that we are just suffixing everything service which I think is one of those dump bucket words like xxxManager etc. In this case the:

- ZuoraSoapService is acting more like a custom Soap client (as @afiore said only query and queryOne are public). 
- SubscriptionService is mainly using ZuoraSoapService to query and is really acting a lot like a repository or similar.   
- MembershipService contains much more of an orchestration/facade logic. 

These are three completely different types of responsibilities with the same suffix. Also theres a lot of stuff going on inside these files that muddle there responsibilities like mapping/transformation code and not so related objects like MemberService containing FrontendMemberRepository a salesforce client from what I can tell. 

Basically where things go and what they should be named is confusing as hell for me

**Comment by tudorraul** [comment]
:+1: bar the comments

**Comment by TesterSpike** [comment]
:+1:  the automated tests have mostly passed (30 out 35) - those that didn't appear to be an issue with a previous change breaking the tests.

File: frontend/app/services/SubscriptionService.scala (hunk 5 above)
Comment (rtyley) [inline frontend/app/services/SubscriptionService.scala]: > a bit unclear why sorting is necessary here, since only one value is then needed?

I've updated the code to try to make it clearer why this is necessary:

https://github.com/guardian/membership-
frontend/blob/fd31fe6aba51/frontend/app/services/SubscriptionService.scala#L83-L113
